### PR TITLE
LDAP integration

### PR DESCRIPTION
**Pull request by fpotier**
- [x] LDAP container
- [x] Initialize the LDAP domain with user
- [x] `backend` config
- [x] `backendnext` config
- [x] `frontend` config
- [x] Global README
- [x] LDAP README
- [x] `backendnext` README
- [x] `backend` README

### CODE REVIEW

File: services/backend/services/v3/config/providers.json [hunk 1]
```diff
@@ -8,5 +8,27 @@
...
+      "displayName": "displayName",
+      "email": "mail"
+    },
+    "server": {
+      "url": "ldap://ldap:389",
+      "bindDn": "cn=admin,dc=facility",
+      "bindCredentials": "admin",
+      "searchBase": "ou=users,dc=facility",
+      "searchFilter": "(uid={{username}})"
+    }
```
[the remark this answers is not in the document]
Comment (minottic) [reply services/backend/services/v3/config/providers.json:32]: I completely agree. Now that we have features toggles, this should all be combinable. I was also thinking, that maybe  one could, instead of mounting the whole config folder in the `fe`, just mount the files which are needed for that specific config and then use the include.path trick, as we do in the be (so there will be a compose.b3.yaml in the fe which has the volume mount of config.v3.json). Maybe this way it will be clearer for the user what it's used? I don't have a strong opinion though...

File: services/backend/services/ldap/config/ldifs/02-users.ldif
Comment (minottic) [inline services/backend/services/ldap/config/ldifs/02-users.ldif:1]: I don't know much of LDAP. Why are these 2 different files?

File: services/backend/services/v4/compose.yaml [hunk 2]
```diff
@@ -2,3 +2,4 @@ include:
   - path:
       - ./compose.base.yaml
       - ./services/elastic/.${ELASTIC_ENABLED:+./../}compose${ELASTIC_ENABLED:+.elastic}.yaml
+      - ../ldap/.${LDAP_ENABLED:+./v4/}compose${LDAP_ENABLED:+.ldap}.yaml
```
Comment (minottic) [inline services/backend/services/v4/compose.yaml:5]: Since LDAP is going to be for both BEs, does it make sense to move this to the backend/compose.yaml?

File: services/frontend/compose.yaml [hunk 3]
```diff
@@ -13,5 +13,6 @@ services:
       - /config/init.sh && nginx -g "daemon off;"
     environment:
       BE_VERSION: ${BE_VERSION:-v4}
+      LDAP_ENABLED: ${LDAP_ENABLED:-}
```
Comment (minottic) [inline services/frontend/compose.yaml:16]: Is the default needed?

File: services/frontend/config/init.sh [hunk 4]
```diff
@@ -3,7 +3,9 @@
 
 apk update && apk add jq
 
-[ "$BE_VERSION" = "v4" ] && EXCLUDE="[!v3]"
+INCLUDE="base"
```
Comment (minottic) [inline services/frontend/config/init.sh:6]: Could this have used GLOBIGNORE (appending all exclusions to it). If we later decide to split the config.json based on functionality, I think excluding will work better

File: services/frontend/config/init.sh
Comment (minottic) [inline services/frontend/config/init.sh:1]: Maybe this refactor could be in a different PR

File: .github/workflows/compose_test.yaml [hunk 5]
```diff
@@ -54,12 +54,13 @@ jobs:
     strategy:
       matrix:
         BE_VERSION: [v3, v4]
-        JOBS_AND_ELASTIC_ENABLED: ['', true]
+        FEATURE_ENABLED: ['', true]
```
Comment (minottic) [inline .github/workflows/compose_test.yaml:57]: It might be good to test the combination of features (or maybe it's too much?). So the matrix could test V3 with jobs and LDAP; V3 with jobs; V3 with LDAP (this means adding a separate LDAP_ENABLED key to the matrix). jobs_and_elastic were merged because they have redundancy once BE_VERSION is set. 

It might be an overkill though

File: services/backend/services/v3/compose.base.yaml [hunk 6]
```diff
@@ -8,7 +8,13 @@ services:
     volumes:
       - ./config/config.local.js:/home/node/app/server/config.local.js
       - ./config/datasources.json:/home/node/app/server/datasources.json
-      - ./config/providers.json:/home/node/app/server/providers.json
+      - ./config/providers.json:/config/providers.base.json
+      - ./config/init.sh:/init.sh
+      - ./config/mergeJson.js:/config/mergeJson.js
+    command:
```
Comment (minottic) [inline services/backend/services/v3/compose.base.yaml:14]: Doesn't this node image have the docker-entrypoint folder?

File: services/backend/services/v3/config/mergeJson.js
Comment (minottic) [inline services/backend/services/v3/config/mergeJson.js:1]: Why couldn't this be the same as the frontend one?

Also, maybe adding the merging logic in the backend could be a separate PR

File: services/backend/services/v3/compose.yaml [hunk 7]
```diff
@@ -2,3 +2,4 @@ include:
   - path:
       - ./compose.base.yaml
       - ./services/rabbitmq/.${JOBS_ENABLED:+./../}compose${JOBS_ENABLED:+.jobs}.yaml
+      - ../ldap/.${LDAP_ENABLED:+./v3/}compose${LDAP_ENABLED:+.ldap}.yaml
```
Comment (minottic) [inline services/backend/services/v3/compose.yaml:5]: See comment on v4/compose.yaml

File: services/backend/services/v3/README.md [hunk 8]
```diff
@@ -75,5 +77,6 @@ Here below we show the internal dependencies of the service, which are not alrea
 graph TD   
     rabbitmq --> archivemock
     rabbitmq --> backend
+    ldap --> backend
```
Comment (minottic) [inline services/backend/services/v3/README.md:80]: Would it make sense, since LDAP is used by both backends, to add a mermaid to the backend/README.md with the services shared by both (it will be useful when adding keycloak)

File: services/frontend/compose.yaml (hunk 3 above)
Comment (fpotier) [reply services/frontend/compose.yaml:16]: Not needed per say, but otherwise you would get a warning

File: services/backend/services/v4/compose.yaml (hunk 2 above)
Comment (fpotier) [reply services/backend/services/v4/compose.yaml:5]: Yes it does!

File: services/backend/services/ldap/config/ldifs/02-users.ldif
Comment (fpotier) [reply services/backend/services/ldap/config/ldifs/02-users.ldif:1]: I did that to make it easier to add new users but I can merge them if you prefer

File: .github/workflows/compose_test.yaml (hunk 5 above)
Comment (fpotier) [reply .github/workflows/compose_test.yaml:57]: I thought about it too but wasn't sure it was needed

File: services/backend/services/ldap/config/ldifs/02-users.ldif
Comment (minottic) [reply services/backend/services/ldap/config/ldifs/02-users.ldif:1]: Ah ok, nice! No I think it's a good idea. Maybe one could add that in the readme then?

File: .github/workflows/compose_test.yaml (hunk 5 above)
Comment (minottic) [reply .github/workflows/compose_test.yaml:57]: It might be useful I think, so we test the multiple merging rules work together. In another pr we can see if it's doable to cache layers in the ci

File: services/frontend/config/init.sh (hunk 4 above)
Comment (fpotier) [reply services/frontend/config/init.sh:6]: Unfortunately, `GLOBIGNORE` is bash only

File: services/backend/services/v3/config/mergeJson.js
Comment (fpotier) [reply services/backend/services/v3/config/mergeJson.js:1]: The container is not running as `root` so I couldn't install `jq` like in the frontend

File: services/backend/services/v3/compose.base.yaml (hunk 6 above)
Comment (fpotier) [reply services/backend/services/v3/compose.base.yaml:14]: No, it only has `docker-entrypoint.sh` 
```
docker run --rm -it ghcr.io/scicatproject/backend-next:v4.5.0 /bin/cat /usr/local/bin/docker-entrypoint.sh
#!/bin/sh
set -e

# Run command with node if the first argument contains a "-" or is not a system command. The last
# part inside the "{}" is a workaround for the following bug in ash/dash:
# https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=874264
if [ "${1#-}" != "${1}" ] || [ -z "$(command -v "${1}")" ] || { [ -f "${1}" ] && ! [ -x "${1}" ]; }; then
  set -- node "$@"
fi

exec "$@"
```

File: services/backend/services/v3/config/mergeJson.js
Comment (minottic) [reply services/backend/services/v3/config/mergeJson.js:1]: what about [this](https://www.npmjs.com/package/node-jq)?

File: .github/workflows/compose_test.yaml [hunk 9]
```diff
@@ -54,13 +54,15 @@ jobs:
     strategy:
       matrix:
         BE_VERSION: [v3, v4]
-        FEATURE_ENABLED: ['', true]
+        JOBS_ENABLED: ['', true]
```
Comment (minottic) [inline .github/workflows/compose_test.yaml:57]: JOBS and ELASTIC can stay as one, as separating them will create redundant combinations

File: services/backend/services/v3/config/mergeJson.js
Comment (fpotier) [reply services/backend/services/v3/config/mergeJson.js:1]: I will try it

File: .github/workflows/compose_test.yaml [hunk 10]
```diff
@@ -54,12 +54,15 @@ jobs:
     strategy:
       matrix:
         BE_VERSION: [v3, v4]
-        JOBS_AND_ELASTIC_ENABLED: ['', true]
+        ELASTIC_OR_JOBS_ENABLED: ['', true]
+        ELASTIC_ENABLED: ['', true]
```
Comment (minottic) [inline .github/workflows/compose_test.yaml:58]: I think `ELASTIC_ENABLED` can be removed

File: services/backend/services/v3/config/mergeJson.js
Comment (minottic) [inline services/backend/services/v3/config/mergeJson.js:1]: this is not needed anymore I think

**Review by minottic — COMMENTED** [review]
I think the `exclude` logic for LDAP is missing from the be init.sh

**Review by minottic — APPROVED** [review]
